src/system/disk_database.rs

```diff
@@ -0,0 +1,445 @@
+                    decode_from_slice(&superblock.data[start..end], bincode::config::standard())
+                        .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))?;
+                decoded.push(value);
+                Ok::<(), io::Error>(())
+            })
+        })?;
+        Ok(decoded)
+    }
+
+    /// Read and decode multiple data from the disk.
+    fn read_multi<T: Decode<()>>(&self, data_infos: Vec<&DataInfo>) -> io::Result<Vec<T>> {
+        if data_infos.is_empty() {
+            return Ok(Vec::new());
+        }
+
+        let mut superblocks = self.split_to_superblocks(data_infos);
+        self.fill_superblocks(superblocks.iter_mut().collect())?;
+        Self::decode_superblocks(superblocks.iter().collect())
+    }
+
+    /// Serializes and writes multiple data to disk. Returns `Vec<DataInfo>` with information about the allocated data.
+    fn write_multi<T: Encode>(&mut self, values: &[&T]) -> io::Result<Vec<DataInfo>> {
+        let encoded = values
+            .iter()
+            .map(|value| encode_to_vec(value, bincode::config::standard()))
+            .collect::<Result<Vec<_>, EncodeError>>()
+            .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))?;
+
+        let encoded_size: usize = encoded.iter().map(|vec| vec.len()).sum();
+        if self.used_size + encoded_size as u64 >= self.total_size {
+            return Err(io::Error::from(io::ErrorKind::OutOfMemory));
+        }
+        let data_infos = encoded.iter().fold(vec![], |mut data_infos, vec| {
+            data_infos.push(DataInfo {
+                offset: self.used_size,
+                data_length: vec.len() as u64,
+            });
+            self.used_size += vec.len() as u64;
+            data_infos
+        });
+
+        let mut encoded = encoded.concat();
+        if self.o_direct {
+            let padding_size = self.padding_to_multiple_block_size(encoded.len() as u64);
+            self.used_size += padding_size;
+            encoded.extend(vec![0; padding_size as usize]); // padding for work with O_DIRECT flag
+        }
+        self.device
+            .write_all_at(&encoded, self.used_size - encoded.len() as u64)?;
+
+        Ok(data_infos)
+    }
+}
+
+impl<K, V> Drop for DiskDatabase<K, V>
+where
+    K: ChunkHash,
+    V: Clone + Encode + Decode<()>,
+{
+    fn drop(&mut self) {
+        if let InitType::RegularFile(file_path) = &self.init_type {
+            std::fs::remove_file(file_path).unwrap()
+        }
+    }
+}
+
+impl<K, V> Database<K, V> for DiskDatabase<K, V>
+where
+    K: ChunkHash,
+    V: Clone + Encode + Decode<()>,
+{
+    fn insert(&mut self, key: K, value: V) -> io::Result<()> {
+        self.insert_multi(vec![(key, value)])
+    }
+
+    fn get(&self, key: &K) -> io::Result<V> {
+        self.get_multi(&[key.clone()]).map(|mut vec| vec.remove(0))
+    }
+
+    fn insert_multi(&mut self, pairs: Vec<(K, V)>) -> io::Result<()> {
+        let mut unique_keys = vec![];
```

Comment:
```suggestion
        let mut unique_keys = Vec::with_capacity(pairs.len());
```
чтобы реаллокаций не было